pkg/reconciler/kubequota/kubequota_controller.go

```diff
@@ -0,0 +1,423 @@
+)
+
+const (
+	controllerName = "kcp-kube-quota"
+)
+
+// Controller manages per-workspace resource quota controllers.
+type Controller struct {
+	queue workqueue.RateLimitingInterface
+
+	resourceQuotaInformer   coreinformers.ResourceQuotaInformer
+	informerFactoryForQuota informerfactory.InformerFactory
+	kubeClusterClient       kubernetes.ClusterInterface
+	informersStarted        <-chan struct{}
+
+	// TODO(ncdc): see if these are interpreted correctly from upstream.
+	// quotaRecalculationPeriod controls how often a full quota recalculation is performed
+	quotaRecalculationPeriod time.Duration
+	// fullResyncPeriod controls how often the dynamic informers do a full resync
+	fullResyncPeriod time.Duration
+
+	workersPerLogicalCluster int
+
+	// lock guards the fields in this group
+	lock                       sync.RWMutex
+	cancelFuncs                map[logicalcluster.Name]func()
+	resourceQuotaEventHandlers map[logicalcluster.Name]cache.ResourceEventHandler
+
+	delegatingEventHandler *delegatingEventHandler
+
+	discoveryStopsLock sync.Mutex
+	discoveryStops     map[logicalcluster.Name]chan struct{}
+
+	// For better testability
+	getClusterWorkspace func(key string) (*tenancyv1alpha1.ClusterWorkspace, error)
+}
+
+// NewController creates a new Controller.
+func NewController(
+	clusterWorkspacesInformer tenancyinformers.ClusterWorkspaceInformer,
+	kubeClusterClient kubernetes.ClusterInterface,
+	resourceQuotaInformer coreinformers.ResourceQuotaInformer,
+	informerFactoryForQuota informerfactory.InformerFactory,
+	quotaRecalculationPeriod time.Duration,
+	fullResyncPeriod time.Duration,
+	workersPerLogicalCluster int,
+	informersStarted <-chan struct{},
+) (*Controller, error) {
+	c := &Controller{
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName),
+
+		resourceQuotaInformer:   resourceQuotaInformer,
+		informerFactoryForQuota: informerFactoryForQuota,
+		kubeClusterClient:       kubeClusterClient,
+		informersStarted:        informersStarted,
+
+		quotaRecalculationPeriod: quotaRecalculationPeriod,
+		fullResyncPeriod:         fullResyncPeriod,
+
+		workersPerLogicalCluster: workersPerLogicalCluster,
+
+		cancelFuncs:                map[logicalcluster.Name]func(){},
+		resourceQuotaEventHandlers: map[logicalcluster.Name]cache.ResourceEventHandler{},
+		delegatingEventHandler: &delegatingEventHandler{
+			eventHandlers: map[schema.GroupResource]map[logicalcluster.Name]cache.ResourceEventHandler{},
+		},
+
+		discoveryStops: map[logicalcluster.Name]chan struct{}{},
+
+		getClusterWorkspace: func(key string) (*tenancyv1alpha1.ClusterWorkspace, error) {
+			return clusterWorkspacesInformer.Lister().Get(key)
+		},
+	}
+
+	clusterWorkspacesInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				c.enqueue(obj)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				c.enqueue(oldObj)
```

Comment:
Why old?

Reply:
Shouldn't matter, I don't think. enqueue just converts it to a key. I can do newObj if that seems better?

Reply:
Key should be the same, so no worries